### PR TITLE
fix(oracle): restart on validator set change

### PR DESCRIPTION
**Pull request by leonz789**
## Description

When node restart, oracle will recache from database to build rounds info for token feeders.
If a validator set change happened within maxNonce blocks, we should not initialize round as `statusOpen`

----

Closes #XXX

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of round statuses and worker management based on new parameters.
	- Improved logic for validator performance evaluation and slashing conditions.

- **Bug Fixes**
	- Adjusted control flow for appending updated feeder IDs during price creation based on transaction context.

- **Tests**
	- Updated test cases to reflect the new method signature for `PrepareRoundEndBlock`.

- **Chores**
	- Updated dependency versions in `go.mod` for improved security and functionality.

### CODE REVIEW

File: x/oracle/keeper/msg_server_create_price.go [hunk 1]
```diff
@@ -70,8 +70,8 @@ func (ms msgServer) CreatePrice(goCtx context.Context, msg *types.MsgCreatePrice
 		)
 		if !ctx.IsCheckTx() {
 			ms.Keeper.GetCaches().RemoveCache(caches)
+			ms.Keeper.AppendUpdatedFeederIDs(msg.FeederID)
 		}
-		ms.Keeper.AppendUpdatedFeederIDs(msg.FeederID)
```
Comment (MaxMustermann2) [inline x/oracle/keeper/msg_server_create_price.go:74]: This looks like it is not backward compatible.

File: x/oracle/keeper/aggregator/context.go [hunk 2]
```diff
@@ -252,10 +252,6 @@ func (agc *AggregatorContext) SealRound(ctx sdk.Context, force bool) (success []
 					}
 					// TODO: optimize operformance
 					sealed = append(sealed, feederID)
-					if !windowClosedMap[feederID] {
-						// this should be clear after performanceReview
-						agc.RemoveWorker(feederID)
-					}
```
Comment (MaxMustermann2) [inline x/oracle/keeper/aggregator/context.go:258]: Please explain this change.

File: x/oracle/module.go [hunk 3]
```diff
@@ -203,155 +203,160 @@ func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.Val
...
-	for _, validator := range validators {
-		power := validatorPowers[validator]
-		reportedInfo, found := am.keeper.GetValidatorReportInfo(ctx, validator)
-		if !found {
-			logger.Error(fmt.Sprintf("Expected report info for validator %s but not found", validator))
-			continue
+	latestValidatorUpdateBlock, ok := am.keeper.GetValidatorUpdateBlock(ctx)
+	p := am.keeper.GetParams(ctx)
+
+	if (!ok || (latestValidatorUpdateBlock.Block <= uint64(ctx.BlockHeight())-uint64(p.MaxNonce))) && len(validatorUpdates) == 0 {
```
Comment (MaxMustermann2) [inline x/oracle/module.go:209]: So slashing can only happen if either
1. there is no last validator update block, or
2. the last update was within the nonce size and there have been no updates in this block.
is that correct?

File: x/oracle/module.go [hunk 4]
```diff
@@ -203,155 +203,160 @@ func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.Val
...
+					logger.Debug(
+						"absent validator",
+						"height", ctx.BlockHeight(),
 						"validator", validator,
+						"missed", reportedInfo.MissedRoundsCounter,
+						"threshold", minReportedPerWindow,
 					)
 				}
+
+				minHeight := reportedInfo.StartHeight + am.keeper.GetReportedRoundsWindow(ctx)
```
Comment (MaxMustermann2) [inline x/oracle/module.go]: set `reportedRoundsWindow := am.keeper.GetReportedRoundsWindow(ctx)` to not call repeatedly.

File: x/oracle/module.go (hunk 3 above)
Comment (leonz789) [reply x/oracle/module.go:209]: 1. no last validator update and current block has no validator set update
2. the last update  was out of the nonce size and there have no updates in current block

File: x/oracle/keeper/aggregator/context.go (hunk 2 above)
Comment (leonz789) [reply x/oracle/keeper/aggregator/context.go:258]: only remove worker when window closed in EndBlock, there's a `defer` to make sure that.

File: x/oracle/module.go (hunk 4 above)
Comment (leonz789) [reply x/oracle/module.go]: updated

File: x/oracle/keeper/aggregator/context.go (hunk 2 above)
Comment (MaxMustermann2) [reply x/oracle/keeper/aggregator/context.go:258]: Got it!

File: x/oracle/keeper/msg_server_create_price.go (hunk 1 above)
Comment (MaxMustermann2) [reply x/oracle/keeper/msg_server_create_price.go:74]: not relevant unless 2/3 of vote power with one node.

File: x/oracle/module.go (hunk 3 above)
Comment (MaxMustermann2) [reply x/oracle/module.go:209]: example to explain:

round size is 3
validator set update is 102
we are at block 103

so 102 <= 103 -3 will fail the check. at this height we don't have history power to do the calculation. hence, skip it.

File: x/oracle/keeper/aggregator/context.go [hunk 5]
```diff
@@ -271,24 +267,25 @@ func (agc *AggregatorContext) SealRound(ctx sdk.Context, force bool) (success []
 
 // PrepareEndBlock is called at EndBlock stage, to prepare the roundInfo for the next block(of input block)
 // func (agc *AggregatorContext) PrepareRoundEndBlock(ctx sdk.Context, block uint64) {
-func (agc *AggregatorContext) PrepareRoundEndBlock(block uint64) (newRoundFeederIDs []uint64) {
+func (agc *AggregatorContext) PrepareRoundEndBlock(block int64, forceSealed bool) (newRoundFeederIDs []uint64) {
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply x/oracle/keeper/aggregator/context.go:270]: Too conservative. Called with `ctx.BlockHeight` which is always positive.

**Review by MaxMustermann2 — APPROVED** [review]

**Review by bwhour — APPROVED** [review]
the whole looks good.

File: x/oracle/keeper/single.go [hunk 6]
```diff
@@ -178,7 +183,7 @@ func initAggregatorContext(ctx sdk.Context, agc *aggregator.AggregatorContext, k
 	// set validatorPower cache
 	c.AddCache(cache.ItemV(validatorPowers))
 
-	agc.PrepareRoundEndBlock(uint64(ctx.BlockHeight()) - 1)
+	agc.PrepareRoundEndBlock(ctx.BlockHeight()-1, false)
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/single.go:186]: int64

**Review by mikebraver — APPROVED** [review]

**Pull request merged**